### PR TITLE
Add providers to look up env vars and system properties by prefix

### LINKED ISSUE #13819: Provide a configuration cache friendly way of looking up a subset of system properties and environment variables
Gradle build configuration cache does not allow a generic way of looking up a set of System properties as build configuration in put. At the elasticsearch build, we currently pass all system properties starting with `tests.` to the test task system properties in a generic way. 
 

### Expected Behavior
Provide a way to use all, or better just a subset of system properties as build configuration input. 

E.g `providers.systemProperties.filter(name -> name.startsWith("test."))`

### Current Behavior

The user can only lookup system properties explicitly one by one by their names. by using `getProviders().systemProperty('name')`. This is not useable in a scenario where I want to use a generic subset of system properties as in put for my build (e.g. everything prefixed with `test.` to pass down to my test tasks. 

### Context

When testing 6.6-rc-1 configuration cache with the elasticsearch master https://github.com/elastic/elasticsearch I stepped into the problem described above when running `./gradlew clean precommit --configuration-cache`

### PR DESCRIPTION
**Pull request by mlopatkin**
There are many user requests to be able to look up a subset of the
environment variables or system properties in configuration-cache
compatible way. Filtering the variables in the build logic makes all
of them, even the unrelated ones, inputs to the configuration cache,
reducing cache hits. The configuration cache machinery is unable to
understand the meaning of the filtering code in the build logic and
,therefore, cannot reduce the number of inputs.
    
This commit adds a specific API to address the most common use case
of selecting a subset of variables with names sharing a common prefix.
When the provider returned by the new API is resolved at the
configuration time, all variables with the prefix become an input.
No other variables affect the cache fingerprint.

Fixes #13819

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle test this

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

@bamboo branching date is close, so I'm going to merge it as is. I'm eager to fix any comments in follow-up PRs.

File: subprojects/model-core/src/main/java/org/gradle/api/internal/provider/sources/SystemPropertiesPrefixedByValueSource.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.stream.Stream;
+
+public abstract class SystemPropertiesPrefixedByValueSource extends MapWithPrefixedKeysValueSource<SystemPropertiesPrefixedByValueSource.Parameters> {
+    public interface Parameters extends MapWithPrefixedKeysValueSource.Parameters {
+    }
+
+    @Override
+    protected Stream<Map.Entry<String, String>> itemsToFilter() {
+        return System.getProperties().entrySet().stream()
+            .filter(e -> e.getKey() instanceof String && e.getValue() instanceof String)
```
[the remark this answers is not in the document]
Comment (mlopatkin) [reply subprojects/model-core/src/main/java/org/gradle/api/internal/provider/sources/SystemPropertiesPrefixedByValueSource.java:29]: Not only that but it is also consistent with `systemProperty` provider which only supports `String` keys and values.

[the remark this answers is not in the document]
Comment (bamboo) [reply subprojects/model-core/src/main/java/org/gradle/api/internal/provider/sources/SystemPropertiesPrefixedByValueSource.java:29]: I see, thanks.

**Pull request merged**